### PR TITLE
[TUTORIAL] Gluon and Sparse NDArray

### PR DESCRIPTION
**Pull request by thomelane**
## Description ##

Added a brand new tutorial on how to use Sparse NDArray with Gluon API. Other tutorials on Sparse NDArray exist but cover low-level details and Module API, not Gluon API. Also shows how to handle data using LibSVM, and why to avoid Dataset and DataLoader. And provides an implementation of a Gluon Block that can be used efficiently with sparse input (unlike nn.Dence).

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by thomelane** [comment]
@eric-haibin-lin @aaronmarkham please review, thanks!

**Comment by anirudhacharya** [comment]
@mxnet-label-bot add [pr-awaiting-review]

File: docs/tutorials/sparse/train_gluon.md [hunk 1]
```diff
@@ -0,0 +1,469 @@
...
+    1.0 97133:0.0024336236529052258 109855:0.9895315766334534 116765:0.2465638816356...
+    0.0 803440:0.4020800292491913
+    
+
+
+Some storage overhead is introduced by serializing the data as characters (with spaces and colons). `dataset.libsvm` is 250 KBs but the original `data` and `label` were 132 KBs combined. Compared with the 4GB dense `NDArray` though, this isn't a huge issue.
+
+### Reading Sparse Data
+
+Using [`LibSVMIter`](https://mxnet.incubator.apache.org/api/python/io/io.html?highlight=libsvmiter), we can quickly and easily load data into batches ready for training. Although Gluon [`Dataset`](https://mxnet.incubator.apache.org/versions/master/api/python/gluon/data.html?highlight=dataset#mxnet.gluon.data.Dataset)s can be written to return sparse arrays, Gluon [`DataLoader`](https://mxnet.incubator.apache.org/versions/master/api/python/gluon/data.html?highlight=dataloader#mxnet.gluon.data.DataLoader)s currently convert each sample to dense before stacking up to create the batch. As a result, [`LibSVMIter`](https://mxnet.incubator.apache.org/api/python/io/io.html?highlight=libsvmiter) is the recommended method of loading sparse data in batches.
```
Comment (aaronmarkham) [inline docs/tutorials/sparse/train_gluon.md]: Have you tested this style of link on beta? I believe I saw that it doesn't parse. Could be a bug on beta...

File: docs/tutorials/sparse/train_gluon.md [hunk 2]
```diff
@@ -0,0 +1,469 @@
...
+                                          init='zeros', allow_deferred_init=True,
+                                          dtype='float32', stype='default', grad_stype='default')
+
+    def hybrid_forward(self, F, x, weight, bias):
+        return F.FullyConnected(x, weight, bias, num_hidden=self._units)
+```
+
+Our `weight` and `bias` parameters are dense (see `stype='default'`) and so are their gradients (see `grad_stype='default'`). Our `weight` parameter has shape `(units, in_units)` because the [`FullyConnected`](https://mxnet.incubator.apache.org/versions/master/api/python/ndarray/ndarray.html?highlight=fullyconnected#mxnet.ndarray.FullyConnected) operator performs the following calculation:
+
+$$Y = XW^T + b$$
```
Comment (aaronmarkham) [inline docs/tutorials/sparse/train_gluon.md]: Is this supposed to render differently on github or just when rendered on the website?

File: docs/tutorials/sparse/train_gluon.md [hunk 3]
```diff
@@ -0,0 +1,469 @@
...
+    print("Memory Allocation for Weight Gradient:")
+    for i in range(len(block_idxs)):
+        print('{:7.3f} MBs ({:7.3f}%) for {:<40}'.format(weight_grad_nbytes[i]/1000000,
+                                                              weight_grad_nbytes_pct[i]*100,
+                                                              blocks[i].name))
+```
+
+### Benchmark: `FullyConnected`
+
+
```
Comment (aaronmarkham) [inline docs/tutorials/sparse/train_gluon.md]: Introduce what's going on in the following code block? Or does the previous go with?

File: docs/tutorials/sparse/train_gluon.md [hunk 4]
```diff
@@ -0,0 +1,469 @@
...
+        loss = loss_fn(pred, label)
+    loss.backward()
+    trainer.step(data.shape[0])
+    loss.wait_to_read()
+```
+
+    532 ms ± 3.47 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
+
+
+![fully_connected](https://raw.githubusercontent.com/dmlc/web-data/master/mxnet/doc/tutorials/ndarray/sparse/fully_connected.png)
```
Comment (aaronmarkham) [inline docs/tutorials/sparse/train_gluon.md]: You can put human text in there instead of underscores if you want. Might be better for screen readers...

File: docs/tutorials/sparse/train_gluon.md [hunk 5]
```diff
@@ -0,0 +1,469 @@
...
+      0.000 MBs (  0.050%) for fullyconnected3                         
+
+
+### Advanced: Sparse `weight`
+
+You can optimize this example further by setting the weight's `stype` to `'row_sparse'`, but whether `'row_sparse'` weights make sense or not will depends on your specific task. 
+
+# Conclusion
+
+As part of this tutorial, we learnt how to write sparse data to disk in LibSVM format and load it back in sparse batches with the [`LibSVMIter`](https://mxnet.incubator.apache.org/api/python/io/io.html?highlight=libsvmiter). We learnt how to improve the performance of Gluon's [`nn.Dense`](https://mxnet.incubator.apache.org/versions/master/api/python/gluon/nn.html?highlight=dense#mxnet.gluon.nn.Dense) on sparse arrays using `mx.nd.sparse`. And lastly, we set `grad_stype` to `'row_sparse'` to reduce the size of the gradient and speed up the parameter update step.
```
Comment (aaronmarkham) [inline docs/tutorials/sparse/train_gluon.md]:
```suggestion
As part of this tutorial, we learned how to write sparse data to disk in LibSVM format and load it back in sparse batches with the [`LibSVMIter`](https://mxnet.incubator.apache.org/api/python/io/io.html?highlight=libsvmiter). We learned how to improve the performance of Gluon's [`nn.Dense`](https://mxnet.incubator.apache.org/versions/master/api/python/gluon/nn.html?highlight=dense#mxnet.gluon.nn.Dense) on sparse arrays using `mx.nd.sparse`. And lastly, we set `grad_stype` to `'row_sparse'` to reduce the size of the gradient and speed up the parameter update step.
```
I can handle the s instead of z and the other mother-tongue 🇬🇧 idiosyncrasies, but `learnt`? Naw, dog. `Learned` just sounds so much better. 🇺🇸

File: docs/tutorials/sparse/train_gluon.md [hunk 6]
```diff
@@ -0,0 +1,469 @@
...
+
+## Recommended Next Steps
+
+* More detail on the [`CSRNDArray`](https://mxnet.incubator.apache.org/api/python/ndarray/sparse.html?highlight=csrndarray#mxnet.ndarray.sparse.CSRNDArray) sparse array format can be found [here]().
```
Comment (aaronmarkham) [inline docs/tutorials/sparse/train_gluon.md]: Switch out "here" with with real descriptive text and a real link....

File: docs/tutorials/sparse/train_gluon.md [hunk 7]
```diff
@@ -0,0 +1,469 @@
...
+* More detail on the [`RowSparseNDArray`](https://mxnet.incubator.apache.org/api/python/ndarray/sparse.html?highlight=rowsparsendarray#mxnet.ndarray.sparse.RowSparseNDArray) sparse array format can be found [here]().
+* Users of the Module API can see a symbolic only example in [this tutorial]().
```
Comment (aaronmarkham) [inline docs/tutorials/sparse/train_gluon.md]: same for "this tutorial"...

File: docs/tutorials/sparse/train_gluon.md [hunk 8]
```diff
@@ -0,0 +1,469 @@
...
+
+# Conclusion
+
+As part of this tutorial, we learnt how to write sparse data to disk in LibSVM format and load it back in sparse batches with the [`LibSVMIter`](https://mxnet.incubator.apache.org/api/python/io/io.html?highlight=libsvmiter). We learnt how to improve the performance of Gluon's [`nn.Dense`](https://mxnet.incubator.apache.org/versions/master/api/python/gluon/nn.html?highlight=dense#mxnet.gluon.nn.Dense) on sparse arrays using `mx.nd.sparse`. And lastly, we set `grad_stype` to `'row_sparse'` to reduce the size of the gradient and speed up the parameter update step.
+
+## Recommended Next Steps
+
+* More detail on the [`CSRNDArray`](https://mxnet.incubator.apache.org/api/python/ndarray/sparse.html?highlight=csrndarray#mxnet.ndarray.sparse.CSRNDArray) sparse array format can be found [here]().
+* More detail on the [`RowSparseNDArray`](https://mxnet.incubator.apache.org/api/python/ndarray/sparse.html?highlight=rowsparsendarray#mxnet.ndarray.sparse.RowSparseNDArray) sparse array format can be found [here]().
```
Comment (aaronmarkham) [inline docs/tutorials/sparse/train_gluon.md]: Switch out "here" with with real descriptive text and a real link....

File: docs/tutorials/sparse/train_gluon.md [hunk 9]
```diff
@@ -0,0 +1,469 @@
...
+      0.000 MBs (  0.000%) for fullyconnected3                         
+    Memory Allocation for Weight Gradient:
+      0.059 MBs ( 93.490%) for fullyconnectedsparse1                   
+      0.004 MBs (  6.460%) for fullyconnected2                         
+      0.000 MBs (  0.050%) for fullyconnected3                         
+
+
+### Advanced: Sparse `weight`
+
+You can optimize this example further by setting the weight's `stype` to `'row_sparse'`, but whether `'row_sparse'` weights make sense or not will depends on your specific task.
```
Comment (eric-haibin-lin) [inline docs/tutorials/sparse/train_gluon.md]: Could you add a link to the contrib.SparseEmbedding https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/gluon/contrib/nn/basic_layers.py#L118 as a reference implementation using sparse weight?

File: docs/tutorials/sparse/train_gluon.md (hunk 1 above)
Comment (thomelane) [reply docs/tutorials/sparse/train_gluon.md]: Works on main website, it doesn't work on beta, but this should be fixed rather than worked around.

File: docs/tutorials/sparse/train_gluon.md (hunk 2 above)
Comment (thomelane) [reply docs/tutorials/sparse/train_gluon.md]: Yep, Github rendering screws up.

File: docs/tutorials/sparse/train_gluon.md (hunk 3 above)
Comment (thomelane) [reply docs/tutorials/sparse/train_gluon.md]: Added a short intro.

File: docs/tutorials/sparse/train_gluon.md (hunk 4 above)
Comment (thomelane) [reply docs/tutorials/sparse/train_gluon.md]: Updated.

File: docs/tutorials/sparse/train_gluon.md (hunk 5 above)
Comment (thomelane) [reply docs/tutorials/sparse/train_gluon.md]: haha, i'll change, but only because you just won the *football*.

File: docs/tutorials/sparse/train_gluon.md (hunk 6 above)
Comment (thomelane) [reply docs/tutorials/sparse/train_gluon.md]: Changed.

File: docs/tutorials/sparse/train_gluon.md (hunk 7 above)
Comment (thomelane) [reply docs/tutorials/sparse/train_gluon.md]: Changed.

File: docs/tutorials/sparse/train_gluon.md (hunk 8 above)
Comment (thomelane) [reply docs/tutorials/sparse/train_gluon.md]: Changed.

File: docs/tutorials/sparse/train_gluon.md (hunk 9 above)
Comment (thomelane) [reply docs/tutorials/sparse/train_gluon.md]: Good example, thanks!

**Comment by thomelane** [comment]
thanks for the reviews @eric-haibin-lin @aaronmarkham. updates made.
